Upgrade to openapi-core 0.13.x and register a view to catch and JSONify openapi-related errors

zupo opened the pull request:
The `openapi-core` library we build on top of is upgraded to `0.13.2`
which brings a complete rewrite of the validation mechanism that is now
based on `jsonschema` library. This manifests as different validation error messages.

Additionally, I moved `openapi_validation_error` from `examples/todoapp` into the main package so it becomes a first-class citizen and people can use it without copy/pasting. If you need custom JSON rendering, you can provide your own `extract_error` function via `pyramid_openapi3_extract_error` config setting.

This is how a migration to this new version looks like in a "realworld app": https://github.com/niteoweb/pyramid-realworld-example-app/pull/90

Closes https://github.com/Pylons/pyramid_openapi3/pull/59
Closes https://github.com/Pylons/pyramid_openapi3/pull/33
Closes https://github.com/Pylons/pyramid_openapi3/pull/43
Closes https://github.com/Pylons/pyramid_openapi3/issues/40
Closes https://github.com/Pylons/pyramid_openapi3/issues/34
Closes https://github.com/Pylons/pyramid_openapi3/issues/35
Refs https://github.com/p1c2u/openapi-core/issues/160

zupo commented:
This is quite a big PR so any extra eyes on it are super helpful! 

cc @gweis, @gotcha, @dz0ny, et al.

zupo commented:
There's a regression though: https://github.com/p1c2u/openapi-core/issues/205

zupo commented:
I've cut a beta release from this PR so you can test it in your projects before I merge and release as stable: https://pypi.org/project/pyramid-openapi3/0.5.0b1/

sayanarijit commented on pyramid_openapi3/exceptions.py:
```diff
@@ -35,5 +34,72 @@ def __str__(self) -> str:
+    """
+    if getattr(err, "schema_errors", None) is not None:
+        for schema_error in err.schema_errors:  # pragma: no branch
+            return extract_error(schema_error)
```
Should it be `return [extract_error(schema_error) for schema_error in err.schema_errors]`?

gweis replied:
most likely yes ... or `yield extract_error(schema_error)` instead of `return`

zupo replied:
No, not really, that makes it wrapped in double list in case of recursion.

gweis replied:
I didn't check what `err.schema_errors` looks like but the code reads like that:
The loop iterates over `err.schema_errors` and returns the first one in this list. In case there are more items in `err.schema_errors` they get ignored.

gweis replied:
Ok, dug around a bit. `schema_errors` seems to be defined only on [InvalidSchemaValue](https://github.com/p1c2u/openapi-core/blob/cb54400136b0e335780dc7eaff769ed3be879115/openapi_core/unmarshalling/schemas/exceptions.py#L30) exceptions.  
I think in this case it really would be interesting to see the full list of errors that caused an `InvalidSchemaValue` error.

sayanarijit replied:
Right. I also think it should return a list instead of dict. I believe other schema parsers also use list. e.g. https://github.com/marshmallow-code/marshmallow/blob/dev/src/marshmallow/exceptions.py#L39

sayanarijit commented on pyramid_openapi3/wrappers.py:
```diff
@@ -1,76 +1,51 @@
+        full_url_pattern = urljoin(request.host_url, path_pattern)
+
+        return OpenAPIRequest(
+            full_url_pattern=full_url_pattern,
```
I think inheriting the `OpenAPIRequest` as a base class and returning `cls(...)` is a better design. So that it can be further extended/inherited.

zupo replied:
I was going with what openapi-core `contib` folder is doing for Django and Flask.

gweis replied:
I think so too. Either extend OpenAPIRequest class or go with a simple module level "factory method" instead of defining a single `@classmethod` on an otherwise unused class.

zupo replied:
I personally prefer the style you both suggested but I think in this case it is better to stick to the same pattern that openapi-core is using, for easier upgrades in the future.

sayanarijit replied:
I went through the https://github.com/p1c2u/openapi-core codebase. Found these classes are (should be named) as `Factory` so that we can easily recognize them as factory classes instead of normal classes.

sayanarijit commented on pyramid_openapi3/wrappers.py:
```diff
@@ -1,76 +1,51 @@
+        cls: t.Type["PyramidOpenAPIResponse"], response: Response
+    ) -> "OpenAPIResponse":
+        """Create OpenAPIResponse from Pyramid Response."""
+        return OpenAPIResponse(
```
Same as `PyramidOpenAPIRequest`.

gweis commented:
Nice Work

1. I would prefer simple factory functions to create `OpenAPIRequest ` and `OpenAPIResponse ` instances. A class with a single `@classmethod` is a bit much for that (unless there are future plans to do more with it). Depending on what the long term plan is, how about adding `aliases` in for the old names and maybe add a deprecation note / warning just in case some one uses these classes directly outside of the lib (which I doubt). 

2. Does it really need the `pyramid_openapi3_JSONify_errors` config option? I am thinking, the `openapi_validation_error` could be registered in any case (it's specific to internal classes). The formatting of exceptions can be customised with `pyramid_openapi3_extract_error` (might be nice to pass in the request as well into that method though). 
Otherwise please call it `pyramid_openapi3_jsonify_errors` :)


Haven't thought that through (and it's probably outside of this PR anyway). I am thinking the `openapi_validation_error` view could inspect / match up the error response in the openapi schema and potentially render the response with the correct mime type, or allow content negotiation to influence the output mime/type. (may that's totally out of scope for this lib, but content negotiation would be nice).

gweis commented:
> There's a regression though: [p1c2u/openapi-core#205](https://github.com/p1c2u/openapi-core/issues/205)

I think I`d prefer getting dicts over unmarshalled classes. But yes should probably highlighted in some release notes or breaking change note or so.

zupo commented:
> I would prefer simple factory functions to create OpenAPIRequest and OpenAPIResponse instances. A class with a single @classmethod is a bit much for that (unless there are future plans to do more with it).

In an ideal world, I would prefer simple factory functions too. But I like to keep the same pattern as openapi-core's `contrib` folder which provides factories for Django and Flask, so it will be easier to keep our factories up-to-date with future changes in openapi-core.

> Depending on what the long term plan is, how about adding aliases in for the old names and maybe add a deprecation note / warning just in case some one uses these classes directly outside of the lib (which I doubt).

Makes sense, but I don't feel strongly about it enough to do it myself. PR welcome!

> Does it really need the pyramid_openapi3_JSONify_errors config option? I am thinking, the openapi_validation_error could be registered in any case (it's specific to internal classes). The formatting of exceptions can already be customised with pyramid_openapi3_extract_error

Good point, I'll do it!

> (might be nice to pass in the request as well into that method though). 

Another good one, will do it!

> Content negotiation for the error response would be nice (maybe another PR someday?)

Yeah, sounds like a great PR.

> Haven't thought that through (and it's probably outside of this PR anyway). I am thinking the openapi_validation_error view could inspect / match up the error response in the openapi schema and potentially render the response with the correct mime type, or allow content negotiation to influence the output mime/type. (may that's totally out of scope for this lib, but content negotiation would be nice).

The current `extract_error` is super rudimentary, i.e. it covers my use cases, and that's about it. I'm sure it's missing things. Hopefully someone someday makes it better.

zupo commented:
Hey @gweis and @sayanarijit, if you can spare a few minutes, I pushed a new commit to address what we discussed above and some additional cleanup/improvements.

sayanarijit commented:
- I think we need to add mypy with typecov in this repo as there are still some `t.Callable` in the code [e.g.](https://github.com/Pylons/pyramid_openapi3/blob/update_to_0_13/pyramid_openapi3/__init__.py).
- For `extract_error`, I think it makes more sense to return a list of errors (and rename to `extract_errors`). Other schema parsers/validators do that. [e.g.](https://github.com/marshmallow-code/marshmallow/blob/dev/src/marshmallow/exceptions.py#L39)
- `PyramidOpenAPI{Request, Response}` should be renamed to `PyramidOpenAPI{Request, Response}Factory` to avoid any confusion that they are supposed to be Factory classes.

zupo commented:
> I think we need to add mypy with typecov in this repo as there are still some t.Callable in the code e.g..

That's already done in master: https://github.com/Pylons/pyramid_openapi3/blob/master/Makefile#L32

> For extract_error, I think it makes more sense to return a list of errors (and rename to extract_errors). Other schema parsers/validators do that. e.g.

Why? The exception returns a list of errors, same as the exception you linked. The `extract_error` is there as an internal helper, not much more. 

> PyramidOpenAPI{Request, Response} should be renamed to PyramidOpenAPI{Request, Response}Factory to avoid any confusion that they are supposed to be Factory classes. e.g.

Correct, will fix.

zupo commented:
Released `0.5.0b2`: https://pypi.org/project/pyramid-openapi3/#history

gweis commented on pyramid_openapi3/exceptions.py:
```diff
@@ -35,5 +35,74 @@ def __str__(self) -> str:
+    """
+    if getattr(err, "schema_errors", None) is not None:
+        for schema_error in err.schema_errors:  # pragma: no branch
+            return extract_error(request, schema_error)
```
I didn't check what err.schema_errors looks like but the code reads like that (and maybe I get a bit fuzzed about it, so please ignore if you disagree)

The loop iterates over `err.schema_errors` and returns the first one in this list. In case there are more items in `err.schema_errors they`` get ignored.

No sure whether this is intended behaviour, but if yes, maybe writing it like this would make the intent obvious?
```
if getattr(err, "schema_errors", None):
  return extract_error(request, err.schema_error[0])
```
should be exactly the same and makes the intention clear?

zupo replied:
Let me look into this one more time.

zupo commented:
Another regression: https://github.com/p1c2u/openapi-core/issues/207

zupo commented:
Released `0.5.0b3`: https://pypi.org/project/pyramid-openapi3/#history

zupo commented:
Hey @gweis and @sayanarijit, can you take a look at https://github.com/Pylons/pyramid_openapi3/pull/60/commits/4a230a0980babfa94c9653204010e3d4a993b7a5 please? I rewrote `extract_error` to support rendering all errors, not just the first one. And added a test to confirm it.

gweis commented:
> Hey @gweis and @sayanarijit, can you take a look at [4a230a0](https://github.com/Pylons/pyramid_openapi3/commit/4a230a0980babfa94c9653204010e3d4a993b7a5) please? I rewrote `extract_error` to support rendering all errors, not just the first one. And added a test to confirm it.

LGTM, thanks.

zupo commented:
Cut a new release 0.5.0b4: https://pypi.org/project/pyramid-openapi3/#history

And here's the test for it: https://github.com/niteoweb/pyramid-realworld-example-app/pull/91

zupo commented:
Deployed to https://app.kafkai.com/api/v1, seems that all systems are green. Ready for merging!
